### PR TITLE
Use git archive instead of a git checkout

### PR DESCRIPTION
**Pull request by ninadpage**
We observed that prereceivecli was calculating incorrect hash when a file from a protected directory was deleted. That turned out to be caused by doing a `checkout` of the incoming revision on top of a `clone` of older/existing revision which leaves the deleted file unstaged in the work tree. This PR is a fix for that.

- According to newer recommendations for pre-receive hooks, you should not update any refs when in the quarantine environment prepared by Git. `checkout` updates the refs.
- Using git archive is an alternative to export the contents of the revision being pushed without updating any refs
- Since now we are only doing strictly read-only operations, we no longer have to leave the quarantine environment by making a deep copy of the repo, so we can clean up a lot of the code.
- Replace ContextManager with a regular class (we have no context to manage anymore)
- Also fixes a small whitespace typo in error message text.

### Impact on Performance

**Current steps:**

- Make a deep copy of the git repo to a temporary directory (which will copy all revs from all branches)
- Do a `git clone` which makes another copy of the entire repo from the temporary directory to `files/.git`
- Do a `git checkout` which makes a copy of one single revision into `files` (work tree)

**Steps with git archive:**

- Do a git archive which makes a copy of one single revision
- Extract the zip archive which makes another copy of one single revision

Since we use `git archive --format=zip` with `-0` [option](https://git-scm.com/docs/git-archive#Documentation/git-archive.txt--0) which disables deflation/compression, there should be no performance penalty associated with zipping/unzipping.

### CODE REVIEW

File: prereceivecli/lib/utils.py [hunk 1]
```diff
@@ -34,19 +34,19 @@
...
+import tempfile
+import zipfile
 from dataclasses import dataclass
+from subprocess import Popen, PIPE, check_output, CalledProcessError
 
 import boto3
 import requests
 from botocore.exceptions import NoRegionError, NoCredentialsError, ClientError
-from commonutilslib import tempdir, Pushd
+from commonutilslib import Hasher
```
Comment (costastf) [inline prereceivecli/lib/utils.py]: Maybe group these imports with a tuple?

**Review by costastf — COMMENTED** [review]
I could really live better without all the whitespace in the functions, if it is all the same to you. Please also add yourself to the contributors file and file headers.

File: prereceivecli/lib/utils.py [hunk 2]
```diff
@@ -209,32 +209,12 @@ def get_table_for_project_group(project_group, credentials):
...
-        self._logger = logging.getLogger(logger_name)
-        self.project = project
-        self.hasher = hasher
-        self.entries = entries
-        self.temporary_directory = None
-        self.git_directory = None
-        self.working_directory = None
+    def __init__(self):
+        self._logger = logging.getLogger(f'{LOGGER_BASENAME}.{self.__class__.__name__}')
+        self.hasher = Hasher()
```
Comment (costastf) [inline prereceivecli/lib/utils.py]: I guess we are only using this in the context of the validation so maybe use it locally there and not have it as an object attribute?

Comment (ninadpage) [reply prereceivecli/lib/utils.py]: Makes sense, done.

**Pull request merged**